### PR TITLE
Support setting padding on the .xterm element

### LINKED ISSUE #946: Add an option to set the terminal padding
I would like to come up with a solution for the following issue:

Say you have a tab that renders the terminal, say it's 300x200. Now if we put the terminal into that tab, we can either make the terminal container to fill it 100%, but in that case the terminal text will touch the edges of the tab at the left, right, top and bottom, which might look weird in some situations. So an easy solution would be to add a 5px margin to the tab container, which does the job, but now you have the viewport scrollbar drawn at a weird position. I think this problem cannot be solved through css (correct me if I'm wrong) because the viewport has to be made aware of the margin to still work in sync. I would therefore propose to make the margin (or better call it padding) configurable through a config option, and then make xterm aware of it:

```javascript
// 5px padding to all sides
term.setOption('padding', 5);

// individual paddings
term.setOption('padding', {
  top: 5, 
  left: 0,
  right: 10,
  bottom: 5
});
```

### The problem in pictures

Fit the tab
clash
5px margin to the tab, weird looking scrollbar
margin bad
5px margin, but scrollbar matching the tab container
margin right

### PR DESCRIPTION
**Pull request by mofux**
Fixes #946 

This PR adds the ability to add a custom css padding to the `.xterm` element while still having the scrollbar filling the whole space. 

**In action:**

### CODE REVIEW

**Comment by blink1073** [comment]
This is great, thanks @mofux!

**Comment by cancan101** [comment]
so what is left to get this merged?

**Comment by ficristo** [comment]
Will this work if I set only padding-left ?

**Comment by mofux** [comment]
@ficristo Yes, it will work with individual paddings for every side.

**Comment by mofux** [comment]
@Tyriar All checks are green now, please review if you find a moment 🧐

**Comment by Tyriar** [comment]
Found a bug, having a non-zero padding breaks IMEs. This is happening because of the position of the xterm-helper and there was an assumption that there was no padding inside the terminal container.

screen shot 2018-01-19 at 12 55 58 pm

kapture 2018-01-19 at 12 54 53

The Clipboard.ts file also moves textarea under the cursor (`moveTextAreaUnderMouseCursor`) for middle click paste selection on Linux and the right click context menu. We'll want to make sure they still work with large padding values.

File: src/Viewport.ts [hunk 1]
```diff
@@ -31,6 +32,10 @@ export class Viewport implements IViewport {
     private scrollArea: HTMLElement,
     private charMeasure: CharMeasure
   ) {
+    // Measure the width of the scrollbar. If it is 0 we can assume it's an OSX overlay scrollbar.
+    // Unfortunately the overlay scrollbar would be hidden underneath the screen element in that case,
+    // therefore we account 15px to make it visible
+    this.scrollBarWidth = (this.viewportElement.offsetWidth - this.scrollArea.offsetWidth) || 15;
```
Comment (Tyriar) [inline src/Viewport.ts]: Let's pull 15 up to the top to a named constant, maybe `FALLBACK_SCROLL_BAR_WIDTH`?

**Comment by cancan101** [comment]
any updates on this PR?

**Comment by mofux** [comment]
@cancan101 I'll try to resolve the remaining issues this week.

**Comment by Tyriar** [comment]
@mofux I synced with master and resolved conflicts then made the following changes:

- Moved `helperContainer` inside `screenElement`, this fixes the issue above where the IME position is incorrect as the helperContainer will always be positioned with the screen elements.
- Removed `screenElement` from the public API, since no one has really asked for it let's not put it in. Consumers also know their own padding so it's fairly easy for them to know the rendered box position.

**Comment by ssh-randy** [comment]
love the change! unfortunately... one small snag.

image

the terminal now bleeds over. i tried setting padding-right, but no dice.

i used this:
```

.xterm-screen {
  padding-left: 20px;
  padding-right: 20px;
  padding-top: 20px;
}
```
any idea how to fix it?

**Comment by ssh-randy** [comment]
ah nevermind. i should be adjusting the css on the container to address this 😄

**Pull request merged**